core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1013 @@
+		let p = Perbill::from_parts(perbill_parts) * int.clone();
+
+		// everything that needs to be either added or subtracted from the original weight.
+		let excess = n.saturating_add(p);
+
+		if positive {
+			int.saturating_add(excess)
+		} else {
+			int.saturating_sub(excess)
+		}
+	}
+}
+
+impl Saturating for Fixed64 {
+	fn saturating_add(self, rhs: Self) -> Self {
+		Self(self.0.saturating_add(rhs.0))
+	}
+	fn saturating_mul(self, rhs: Self) -> Self {
+		Self(self.0.saturating_mul(rhs.0) / DIV)
+	}
+	fn saturating_sub(self, rhs: Self) -> Self {
+		Self(self.0.saturating_sub(rhs.0))
+	}
+}
+
+/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
+/// for safe addition.
+impl ops::Add for Fixed64 {
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self::Output {
+		Self(self.0 + rhs.0)
+	}
+}
+
+/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
+/// for safe subtraction.
+impl ops::Sub for Fixed64 {
+	type Output = Self;
+
+	fn sub(self, rhs: Self) -> Self::Output {
+		Self(self.0 - rhs.0)
+	}
+}
+
+impl CheckedSub for Fixed64 {
+	fn checked_sub(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_sub(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+impl CheckedAdd for Fixed64 {
+	fn checked_add(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_add(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+/// Helper gcd function used in Rational128 implementation.
+fn gcd(a: u128, b: u128) -> u128 {
+	match ((a, b), (a & 1, b & 1)) {
+		((x, y), _) if x == y => y,
+		((0, x), _) | ((x, 0), _) => x,
+		((x, y), (0, 1)) | ((y, x), (1, 0)) => gcd(x >> 1, y),
+		((x, y), (0, 0)) => gcd(x >> 1, y >> 1) << 1,
+		((x, y), (1, 1)) => {
+			let (x, y) = (min(x, y), max(x, y));
+			gcd((y - x) >> 1, x)
+		},
+		_ => unreachable!(),
+    }
+}
+
+/// A wrapper for any rational number with a 128 bit numerator and denominator.
+#[derive(Clone, Copy, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Rational128(u128, u128);
+
+/// Some helper functions to work with 128bit numbers. Note that the functionality provided here is
+/// only sensible to use with 128bit numbers because for smaller sizes, you can always rely on
+/// assumptions of a bigger type (u128) being available, or simply create a per-thing and use the
+/// multiplication implementation provided there.
+pub mod helpers_128bit {
+	/// return the number of bits that are needed to store `n`.
+	pub(crate) fn bits_of(n: u128) -> u32 {
+		128 - n.leading_zeros()
+	}
+
+	/// Tries to compute `a * b / c`. The approach is:
+	///   - Simply try a * b / c.
+	///   - Swap the operations in case the former multiplication overflows. Divide first. This
+	///     might collapse to zero.
+	///
+	/// If none worked then return Error.
+	pub fn multiply_by_rational(a: u128, b: u128, c: u128) -> Result<u128, &'static str> {
```

Comment:
isn't the second approach a bit too much unprecise in some case like this one 
`assert_eq!(multiply_by_rational(2*MAX64 - 1, MAX64, MAX64).unwrap(), 2*MAX64 - 1);`

Reply:
Yeah, frankly, this is horrible. I will rethink it.

Though since we don't really use it directly. If we couldn't come up with some other approach, the function can/should be removed altogether. Then `multiply_by_rational_greedy()` just does the first part of this function and falls back to the greedy approach if it fails.

Reply:
This is equal to computing:

```
// ideally 
31 * 15  / 16 ~= 30
// but gives 
31 * 15  / 16 ~= 15
```

in an 8-bit number scale and I don't see any other trivial way to do it. You multiply first, you overflow. you divide first, either you will collapse to zero (15 / 16) or you will lose a lot of accuracies (31 / 16)

Swapping is useful when `a` is many orders of magnitude bigger than `b` and `c`. But I couldn't really formulate how it can be used safely.. hence: 
- remove the swap
- if `b < c` then we build a perQuintill and use that. 
- otherwise, we don't do anything and return an error. 

This makes some tests that were particularly easy and accurate with the swapping method to become a bit inaccurate, but less horrible

```rust
assert_eq!(
	multiply_by_rational_greedy(MAX128, 2, 3),
	226854911280625642 082061493673886498661,
        // with swap this would accurately be computed to: MAX128 / 3 * 2
        // which is
     // 226854911280625642 308916404954512140970
);
``` 

which is still inaccurate but better I suppose.

cc @dvdplm

Reply:
regarding use case: no, it is also used in the inner implementation of `Rational128` for addition etc. 

regarding providing something minimal that works rather than something flaky: I agree. I've changed the way that both functions work together now and they actually cover all the edge cases that we discussed here. I think it is good to be used for now. Regardless, what we were using before was horribly simple (naively trying to swap the operations and that was it), so this PR is a step forward anyhow. If we really need the multi-limb arithmetic, that is a bigger task and we can work on it in a follow-up PR.